xffind/fgr/man/fgr.1
.\" SCCS ID: fgr.1 12/10/2000
.TH fgr 1 "http://rodent.xffm.org"
.SH NAME
fgr \- File content search command
.SH SYNOPSYS
.B fgr 
[-r] [-v] [-d ddd] [-m mmm] [-f filter] [-s (+/-)size]
[-t type] [-p perm] [grep options...] 
.PP
.TP 6
.BI \-v 
.br
verbose
.TP
.BI \-V 
.br
Print version number information
.TP
.BI \-P 
.br
print process id
.TP
.BI \-f\ filter
.br
file filter (enclosed in quotes if regexp *,? or [] is used)
.TP
.BI \-r 
.br
recursive
.TP
.BI \-s\ \+kbytes
.br
Size greater than kbytes KBYTES
.TP
.BI \-s\ \-kbytes
.br
size less than kbytes KBYTES
.TP
.BI \-p\ perm
.br
perm is either suid | exe
.TP
.BI \-t\ type
.br
reg | dir | sym | sock | blk | chr | fifo (regular, directory, symlink, socket, blk_dev, 
chr_dev, fifo)
.TP
.BI \-d\ ddd
.br
created or modified in the previous (int) dd days
.TP
.BI \-m\ mmm
.br
created or modified in the previous (int) mm months
.TP
.BI \-p\ perm
.br
perm is either suid | exe
.TP
.BI \-e\ string
.br
containing string (if *,? or [], use quotes)
.TP
.BI \-E\ regexp
.br
containing regexp: (use quotes).
.TP
.BI \-i 
.br
ignore case (for search string -c)
.TP
.BI \-y 
.br
same as -i (obsolete)
.TP
.BI \-L 
.br
print the  name  of each input file from which *no* 
output would normally have been printed.
.TP
.BI \-c 
.br
only print a count of matching lines for each input file.
.TP
.BI \-w 
.br
Select  only  those  lines  containing matches that 
form whole words. Word-constituent  characters  are 
letters, digits, and the underscore.
.TP
.BI \-x 
.br
Select only those matches that  exactly  match  the whole line.
.TP
.BI \-Z 
.br
Output  a  zero  byte  (the  ASCII  NUL  character) 
instead of the character that  normally  follows  a 
file  name
.PP
.SH DESCRIPTION
fgr is a file content search engine for Rodent.
.PP
.SH REQUIREMENTS
.B fgr requires GNU grep 2.x or higher
.SH SEE ALSO

.PP
.SH AUTHOR
Edscott Wilson Garcia <edscott@imp.mx>



ChangeLog
* commit fca2ccc01f90b2589d3d1cf2776b7fb16a6fbb70--------------------------
Date:   Thu Aug 25 16:57:30 2016 -0500

    Csh completion almost done
    
    Debug pending. Completion suggestion does not emerge yet.

* commit a9ba157801182ecb32e60807b1879fe3fd2ab4ea--------------------------
Date:   Thu Aug 25 14:00:07 2016 -0500

    bash completion fully functional now

* commit 8b3e43a07f8f7fc0297da24a8a86d4b8fd374af7--------------------------
Date:   Wed Aug 24 20:34:47 2016 -0500

    print_c class nearly complete
    
    only show_text method not passed test.

* commit 7a54df23b6a65e7a90073cf9849f196542230cdb--------------------------
Date:   Wed Aug 24 17:08:46 2016 -0500

    Enabled initial print command
    
    pending:
      print_error
      print_debug
      print_icon
      tag testing

* commit fa9ef5de9808610752815d53018df227c762e627--------------------------
Date:   Tue Aug 23 20:48:48 2016 -0500

    update ChangeLog

* commit 089c7465aac47f490043f3d4f080db7b7c8e8ed9--------------------------
Date:   Tue Aug 23 20:47:27 2016 -0500

    Switch focus from iconview to status on demand
    
    This solves the issue of the nonvisible cursor.

* commit a67839df99840c5204129e7ebd04a26bff77f420--------------------------
Date:   Tue Aug 23 17:01:26 2016 -0500

    Tweaks on the keyboard input and associated widgets
    
    - cursor only appears in status if clicked. Need to try 3.20
      instruction or generate the click effect (focus?)

* commit 70e211e8a123b366dd2c78dc80776570a0b6f6ea--------------------------
Date:   Tue Aug 23 13:30:49 2016 -0500

    Small tweaks

* commit 9e9205a3bfe6a62e2e253ec78bae88bff871d03f--------------------------
Date:   Tue Aug 23 12:01:07 2016 -0500

    key processing now in working order for lpterm
    
    * Nothing done with keys pressed, as of this commit.

* commit 75a1dc8ec15f6a41101bff47b34a96e8db63844f--------------------------
Date:   Mon Aug 22 20:50:13 2016 -0500

    Add more lpterm_c class code
    
    * this commit does not compile. Code yet to be cleansed and optimized
      for class structure.

* commit df2fafaf7b1348bd135a70b3367d260cc200c2e6--------------------------
Date:   Mon Aug 22 16:52:54 2016 -0500

    Start introduction of lpterm code
    
    Now testing new keypress callback route

* commit 91a7ed62781d5eae5e47e340b7a298b2cebe0655--------------------------
Date:   Mon Aug 22 13:28:51 2016 -0500

    Update ChangeLog

* commit bd889cd2f228e495e7ef2e942634354f2745fb43--------------------------
Date:   Mon Aug 22 13:28:06 2016 -0500

    Pathbar debugged and working fine.

* commit fdeeb6b32874693c63c917f0d6e7346a29c47433--------------------------
Date:   Mon Aug 22 11:09:08 2016 -0500

    Taskbar now enabled
    
    * compiles ok, but will lead to crash. Need to debug it.

* commit b0166d4dd64d8bab663f6a179eeb5454492503bb--------------------------
Date:   Sun Aug 21 21:14:54 2016 -0500

    Added pathbar_c class code
    
    * not yet finished nor compiled.

* commit aaafef0b3184a506d35e7fdd8f2ab078df70181d--------------------------
Date:   Sun Aug 21 20:23:58 2016 -0500

    Added callback for the go-home button
    
    * compiled and tested fine.

* commit bc0aa606743a979d98278811624f2acd81c72178--------------------------
Date:   Sun Aug 21 19:00:06 2016 -0500

    More iconview improvements
    
    * icon now updates to tab label
    * non directory highlight with a composite NW-folder-open icon
    * gtk_c object shared from window_c object to all others (to share
      pixbuf hash)

* commit 924f653e5b97d91ddc214c4a1aa1cb9836cb5dba--------------------------
Date:   Sat Aug 20 21:19:08 2016 -0500

    Class optimizations and increased icon resolution
    
    * Application icon now set
    * Conventional icons in homedir now set (Bookmarks, Downloads, Music, etc)
    * view_c creation method simplified.

* commit 33dfa7332852a4628526f1d779e24fb7f55006e8--------------------------
Date:   Fri Aug 19 14:57:56 2016 -0500

    Added initial code for mime-module
    
    * should simplify this into a class.

* commit 8de79ef9aa183194cfe4c58bc3a8053a219777ce--------------------------
Date:   Fri Aug 19 12:18:12 2016 -0500

    Installation now finishes correctly defined.

* commit e478b3ad528c7d83f81642cd7b2a54bd2c66e184--------------------------
Date:   Fri Aug 19 11:17:27 2016 -0500

    Add more installation files

* commit 6797d5b8835b15ef80a231bf07e622acad6eeaab--------------------------
Date:   Fri Aug 19 10:03:08 2016 -0500

    Added files for configure script

* commit 38bd03d24f270588672aef2cefce9ae09c1cf549--------------------------
Date:   Thu Aug 18 21:34:58 2016 -0500

    Avoid highlight in large directories
    
     highlight hogs cpu in large directories. Just as no stat nor composite icons
     in directories with more than 500 elements, no highlighting is done either.

* commit a395848981ceee96f3cddc8ce214485f1050d02e--------------------------
Date:   Thu Aug 18 21:19:56 2016 -0500

    Finished details for icon emblems
    
    * Bug detected: cpu hog on large directories (/usr/bin)

* commit 95adf7e46e10a99988d8cd74a882625f029ebb14--------------------------
Date:   Thu Aug 18 16:57:08 2016 -0500

    Tweaks on emblems for icons with stat
    
    final checks pending

* commit a11149ddf4d7f6042ade8d82493ab85d846dd3f7--------------------------
Date:   Thu Aug 18 12:15:46 2016 -0500

    Added composite icon code with cairo class
    
    * not yet tested. Still need to generate emblem svg icons.

* commit e8017f870c7ea4611e670f2a67764a1fcd398588--------------------------
Date:   Wed Aug 17 16:55:48 2016 -0500

    icon highlighting
    
    next in highlighting will be to use dt_type to determine icon type for
    highlighting . Document new for documents and execute for executables.

* commit 8ab73be2765662c19dd3fcf2b388e8ced91420e2--------------------------
Date:   Tue Aug 16 21:14:15 2016 -0500

    Icon highlighting
    

xffind/fgr/src/fgr.c

/* glob.c file filter for grep.*/
/* 
   Copyright 2000-2010 Edscott Wilson Garcia

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program;  */

/*****************************************************************/

#include "fgr-globber.h"

/** tripas */
#define GREP "grep"

static void *object = NULL;
static int initial;
static int terminated = 0;
static char *token;
static int options = 0,
    type = 0;
static long size = 0;
static long month_t = 0;
static long unsigned day_t = 0;
static long unsigned hour_t = 0;
static long unsigned min_t = 0;

/* two low order bytes are defined in globber.h */
#define GLOBRUN_PID     	0x10000
#define GLOBRUN_COUNT   	0x20000
#define GLOBRUN_FILTERED   	0x40000
#define GLOBRUN_IGNORE_CASE	0x80000
#define GLOBRUN_REG_EXP  	0x100000
#define GLOBRUN_INVERT       	0x200000
#define GLOBRUN_WORDS_ONLY   	0x400000
#define GLOBRUN_LINES_ONLY   	0x800000
#define GLOBRUN_ZERO_BYTE    	0x1000000
#define GLOBRUN_NOBINARIES   	0x2000000

#define MAX_ARG 25
#ifndef COPYRIGHT
# define COPYRIGHT "Copyright (c) 2002-2010 Edscott Wilson Garcia. GPL distribution licence."
#endif

static int
display (char *input) {
    if(terminated)
        return terminated;      /* die quietly and quickly */
    printf ("%s\n", input);
    if(time (NULL) - initial > 3) {
        fflush (NULL);
        initial = time (NULL);
    }
    return terminated;
}

static int
grep (char *file) {
    static char *arguments[MAX_ARG];
    int status = 0;
    if(terminated)
        return terminated;      /* die quietly and quickly */

    arguments[status++] = "grep";
#ifdef GNU_GREP
    //arguments[status++] = "-d";
    //arguments[status++] = "skip";
    //arguments[status++] = "-H";
    if(options & GLOBRUN_NOBINARIES)
        arguments[status++] = "-I";
    if(options & GLOBRUN_ZERO_BYTE)
        arguments[status++] = "-Z";
    if(options & GLOBRUN_INVERT) {
        arguments[status++] = "-L";
    } else {
        arguments[status++] = "-l";
    }
#else
    if(options & GLOBRUN_INVERT) {
        arguments[status++] = "-v";
    } else {
        arguments[status++] = "-l";
    }
#endif
    if(options & GLOBRUN_IGNORE_CASE)
        arguments[status++] = "-i";
    if(options & GLOBRUN_WORDS_ONLY)
        arguments[status++] = "-w";
    if(options & GLOBRUN_LINES_ONLY)
        arguments[status++] = "-x";

    if((options & GLOBRUN_COUNT)) {
        arguments[status++] = "-c";
    }
#ifdef GNU_GREP
    if(options & GLOBRUN_REG_EXP)
        arguments[status++] = "-E";
    else
#endif
        arguments[status++] = "-e";
    
    arguments[status++] = token;

    arguments[status++] = file;
    arguments[status++] = (char *)0;
#if 0
  //  if (options & GLOBBER_VERBOSE)
  {
    int i;
    for (i = 0; arguments[i] != NULL; i++)
      fprintf(stderr,"%s ", arguments[i]);
    fprintf(stderr,"\n");
  }
#endif
    if(fork () == 0) {
        execvp (GREP, arguments);
        fprintf(stderr, "%s: %s\n", strerror (ENOENT), GREP);
        exit (1);
    }
    wait (&status);

    return terminated;
}

static char *message[] = {
    " ",
    N_("Options:")," ","[-vVPrMACaiIyLcwxZ] [-fpotkhsmudgeE]] ["N_("Path"),"]", "\n", " \n",
    "fgr: ", N_("Locate documents and folders on this computer by name or content"), "\n", " \n",
    N_("Filter:"), N_("Search for files"),"\n",
    //N_("-k | -h | -d | -m --->  -M | -C | -A"), "\n",
    N_("Contents:"), N_("File's contents filtered as plain text."), "\n", " \n",

    "-v", "\t\t",
    N_("Verbose output debugging mode."), "\n",
    "-V", "\t\t",
    N_("Show version information"), "\n"," \n",

    N_("Filter Details"),"\n"," \n",

    "-a", "\t\t",
    N_("Stay on single filesystem"),"\n",
    "-P", "\t\t",
    N_("Print...") ," ", N_("The unique Process ID that identifies this process."),"\n",
    "-D", "\t\t",
    N_("Find hidden files and directories"), "\n",
    "-f ", N_("Filter"),"\t"
    N_("To display files which pass the filter (stars and regexps will be processed)"),"\n",
    "-r", "\t\t",
    N_("Recursive"),"\n",
    "-s ", "+",N_("KByte"), "\t",
    N_("is greater than or equal to"), " (KB): ", N_("KByte"),"\n",
    "-s ", "-", N_("KByte"), "\t",
    N_("is less than or equal to"), " (KB): ", N_("KByte"),"\n",
    "-p ", "suid | exe", "\t",
    N_("Set UID"), "(suid) | ", N_("Executable Text Files"), " (exe)\n",
    "-o ", "octal", "\t",
    N_("The permissions of the file, in octal notation."),"\n",
    "-t ", N_("Type"), " \t",
    " any | reg | dir | sym | sock | blk | chr | fifo","\n",
    "\t\t",N_("Any")," (any) ", N_("Default type"),"\n",
    "\t\t",N_("Regular file")," (reg)\n",
    "\t\t",N_("Directory")," (dir)\n",
    "\t\t",N_("Symbolic Link")," (sym)\n",
    "\t\t",N_("UNIX Socket")," (sock)\n",
    "\t\t",N_("Block device")," (blk)\n",
    "\t\t",N_("Character device")," (chr)\n",
    "\t\t",N_("FIFO")," (fifo)\n",

    "-k ", N_("Minutes"),"\t",
    N_("Details"),": ", N_("TIME"),
      " (", N_("Previous"), "): " N_("Minutes"), " [--> -M | -C | -A]", "\n",
    "-h ", N_("Hours"), "\t",
    N_("Details"),": ", N_("TIME"),
      " (", N_("Previous"), "): " N_("Hours"), " [--> -M | -C | -A]", "\n",
    "-d ", N_("Days"), " \t",
    N_("Details"),": ", N_("TIME"),
      " (", N_("Previous"), "): " N_("Days"), " [--> -M | -C | -A]", "\n",
    "-m ", N_("Month"), "\t", 
    N_("Details"),": ", N_("TIME"),
      " (", N_("Previous"), "): " N_("Month(s)"), " [--> -M | -C | -A]", "\n",

    "-M ", N_("TIME"), "\t",
    N_("Modification Time :"),  N_("TIME"), " (", 
	"mknod, truncate, utime, write)", "\n", 

    "-A ", N_("TIME"), "\t",
    N_("Access Time :"),  N_("TIME"), " (",
	"exec, mknod, pipe, utime, read)", "\n", 

    "-C ", N_("TIME"), "\t",
    N_("Creation Time :"),  N_("TIME"), " (",
	"chown, chgrp, chmod, ln)", "\n", 

    "-u ", N_("uid"), "  \t",
    N_("User:"), " ", N_("uid"),"\n",

    "-g ", " ", N_("gid"), "  \t",
    N_("Group:"), " ",  N_("gid"),"\n",

    "-Z", "\t\t",
    N_("Output  a  zero  byte  (the  ASCII  NULL  character) instead of the character that  normally  follows  a file  name"),"\n \n",
    
    N_("Content View"), ": --------\n \n", 
    "-e ", N_("STRING"), "\t"
    N_("Regular expression"), ": ", N_("STRING"), "\n",
    "-E ", N_("STRING"), "\t",
    N_("Regular expression"), "(", N_("Extended"), "): ", N_("STRING"), "\n",
    "-i", "\t\t",
    N_("Ignore case"), "\n",
    "-I", "\t\t",
    N_("Text Files"), "\n", 
    "-L", "\t\t",
    N_("No match"), "\n", 
    "-c", "\t\t",
    N_("Line Count"), "\n",
    "-w", "\t\t",
    N_("Whole words only"), "\n",
    "-x", "\t\t",
    N_("Match whole lines only"), "\n \n ",
    "fgr-", VERSION, " ", COPYRIGHT, "\n",
    NULL
};

void
finish (int sig) {
    /*printf("\n****\nglob terminated by signal\n****\n"); */
    terminated = 1;
    fflush (NULL);
}

void
halt (int sig) {
    fflush (NULL);
    globber_destroy (object);
    exit (1);
}

#define CHECK_ARG if (argc <= i) goto error;
int
main (int argc, char **argv) {
    int i,
      timetype = 0;
    char *filter = "*";
    int (*operate) (char *) = display;
    char *path = ".";
    initial = time (NULL);

#ifdef ENABLE_NLS
    setlocale (LC_MESSAGES, "");
    setlocale (LC_ALL, "");
    bindtextdomain (PACKAGE, PACKAGE_LOCALE_DIR);
    textdomain (PACKAGE);
#endif
    /* initializations  */
    signal (SIGHUP, halt);
    signal (SIGSEGV, finish);
    signal (SIGKILL, finish);
    signal (SIGTERM, finish);

    if(argc < 2) {
      error:
        fprintf (stdout, _("Usage : %s [OPTIONS]\n"), argv[0]);
        i = 0;
        while(message[i])
            fprintf (stdout, "%s", _(message[i++]));
        exit (1);
    }
    object = globber_create ();
    for(i = 1; i < argc; i++) {
        if(argv[i][0] == '-') {
            /* options for the globber : **************** */
            if(strstr (argv[i], "v") != NULL) {
                glob_set_options (object, GLOBBER_VERBOSE);
                options |= GLOBBER_VERBOSE;
                continue;
            }
            if(strstr (argv[i], "M") != NULL) {
                timetype = 1;
                glob_set_options (object, GLOBBER_MTIME);
                continue;
            }
            if(strstr (argv[i], "A") != NULL) {
                timetype = 1;
                glob_set_options (object, GLOBBER_ATIME);
                continue;
            }
            if(strstr (argv[i], "C") != NULL) {
                timetype = 1;
                glob_set_options (object, GLOBBER_CTIME);
                continue;
            }
            if(strstr (argv[i], "a") != NULL) {
                glob_set_options (object, GLOBBER_XDEV);
                options |= GLOBBER_XDEV;
                continue;
            }
            if(strstr (argv[i], "v") != NULL) {
                glob_set_options (object, GLOBBER_VERBOSE);
                options |= GLOBBER_VERBOSE;
                continue;
            }
            if(strstr (argv[i], "r") != NULL) {
                glob_set_options (object, GLOBBER_RECURSIVE);
                options |= GLOBBER_RECURSIVE;
                continue;
            }
            if(strstr (argv[i], "R") != NULL) {
                glob_set_options (object, GLOBBER_RECURSIVE);
                glob_set_options (object, GLOBBER_RECURSIVE_NO_HIDDEN);
                options |= GLOBBER_RECURSIVE;
                options |= GLOBBER_RECURSIVE_NO_HIDDEN;
                continue;
            }
            if(strstr (argv[i], "u") != NULL) {
                i++;
                CHECK_ARG;
                glob_set_user (object, atol (argv[i]));
                continue;
            }
            if(strstr (argv[i], "g") != NULL) {
                i++;
                CHECK_ARG;
                glob_set_group (object, atol (argv[i]));
                continue;
            }

            if(strstr (argv[i], "t") != NULL) {
                i++;
                type &= 07777;
                CHECK_ARG;
                /*if (strcmp (argv[i], "any") == 0) type &= 07777; */
                if(strcmp (argv[i], "reg") == 0)
                    type |= S_IFREG;
                if(strcmp (argv[i], "dir") == 0)
                    type |= S_IFDIR;
                if(strcmp (argv[i], "sym") == 0)
                    type |= S_IFLNK;
                if(strcmp (argv[i], "sock") == 0)
                    type |= S_IFSOCK;
                if(strcmp (argv[i], "blk") == 0)
                    type |= S_IFBLK;
                if(strcmp (argv[i], "chr") == 0)
                    type |= S_IFCHR;
                if(strcmp (argv[i], "fifo") == 0)
                    type |= S_IFIFO;
                if(strcmp (argv[i], "any") != 0) {
                    glob_set_options (object, GLOBBER_TYPE);
                    glob_set_type (object, type);
                }
                continue;
            }
            if(strstr (argv[i], "p") != NULL) {
                i++;
                /*type &= S_IFMT; */
                CHECK_ARG;
                if(strcmp (argv[i], "suid") == 0)
                    type |= S_ISUID;
                if(strcmp (argv[i], "exe") == 0)
                    type |= S_IXUSR;
                glob_set_options (object, GLOBBER_PERM);
                glob_set_type (object, type);
                continue;
            }
            if(strstr (argv[i], "o") != NULL) {
                int valor;
                i++;
                type &= S_IFMT;
                CHECK_ARG;
                sscanf (argv[i], "%o", &valor);
                type |= (07777 & valor);
                glob_set_options (object, GLOBBER_PERM);
                glob_set_type (object, type);
                continue;
            }

            if(strstr (argv[i], "s") != NULL) {
                i++;
                CHECK_ARG;
                size = atol (argv[i]);
                if(size < 0){
                    glob_set_sizeL (object, (-1) * size * 1000);
		} else {
                    glob_set_sizeG (object, size * 1000);
		}
                continue;
            }

            if(strstr (argv[i], "D") != NULL) {
                glob_set_options (object, GLOBBER_ADD_DOT_FILTER);
                continue;
            }
	    // time
	    // minutes
	    if(strstr (argv[i], "k") != NULL) {
                i++;
                CHECK_ARG;
                glob_set_options (object, GLOBBER_TIME);
                min_t = atol (argv[i]);
                glob_set_minutes (object, min_t);
                continue;
            }
	    // hours
            if(strstr (argv[i], "h") != NULL) {
                i++;
                CHECK_ARG;
                glob_set_options (object, GLOBBER_TIME);
                hour_t = atol (argv[i]);
                glob_set_hours (object, hour_t);
                continue;
            }
	    // days
            if(strstr (argv[i], "d") != NULL) {
                i++;
                CHECK_ARG;
                glob_set_options (object, GLOBBER_TIME);
                day_t = atol (argv[i]);
                glob_set_days (object, day_t);
                continue;
            }
	    // months
            if(strstr (argv[i], "m") != NULL) {
                i++;
                CHECK_ARG;
                glob_set_options (object, GLOBBER_TIME);
                month_t = atol (argv[i]);
                glob_set_months (object, month_t);
                continue;
            }

            if(strstr (argv[i], "f") != NULL) {
                options |= GLOBRUN_FILTERED;
                i++;
                CHECK_ARG;
                filter = argv[i];
                /*if (options & GLOBBER_VERBOSE) 
                 *  fprintf(stderr, "filtering %s\n", filter);*/
                continue;
            }
            /* options for grep : ****************** */
            if(strstr (argv[i], "I") != NULL) {
                options |= GLOBRUN_NOBINARIES;
                continue;
            }
            if((strstr (argv[i], "i") != NULL) || (strstr (argv[i], "y") != NULL)) {
                options |= GLOBRUN_IGNORE_CASE;
                continue;
            }
            if(strstr (argv[i], "L") != NULL) {
                options |= GLOBRUN_INVERT;
                continue;
            }
            if(strstr (argv[i], "c") != NULL) {
                options |= GLOBRUN_COUNT;
                continue;
            }
            if(strstr (argv[i], "w") != NULL) {
                options |= GLOBRUN_WORDS_ONLY;
                continue;
            }
            if(strstr (argv[i], "x") != NULL) {
                options |= GLOBRUN_LINES_ONLY;
                continue;
            }
            if(strstr (argv[i], "Z") != NULL) {
                options |= GLOBRUN_ZERO_BYTE;
                continue;
            }
            if(strstr (argv[i], "P") != NULL) {
                options |= GLOBRUN_PID;
                printf ("PID=%d\n", (int)getpid ());
                fflush (NULL);
                continue;
            }
            if(strstr (argv[i], "E") != NULL) {
                i++;
                CHECK_ARG;
                token = argv[i];
                operate = grep;
                options |= GLOBRUN_REG_EXP;
                continue;
            }
            if(strstr (argv[i], "e") != NULL) {
                i++;
                CHECK_ARG;
                token = argv[i];
                operate = grep;
                options |= GLOBRUN_REG_EXP;
                options ^= GLOBRUN_REG_EXP;     /* turn off extended regexp */
                continue;
            }

            if(strstr (argv[i], "V") != NULL) {
                printf ("This is %s %s (find<-->grep)\nForward any bug reports to %s\n%s\n", "Rodent fgr", PACKAGE_VERSION,
                        PACKAGE_BUGREPORT, COPYRIGHT);
                return 0;
            }
            fprintf (stdout, "Oops... %s: %s\n", strerror (EINVAL), argv[i]);
            exit (1);
        }
        if(((min_t) || (hour_t) || (day_t) || (month_t)) && !timetype)
            glob_set_options (object, GLOBBER_MTIME);
        if(argv[i] != NULL)
            path = argv[i];
    }                           /* end of argument processing */

    terminated = globber (object, path, operate, filter);
/*	if (terminated) printf("glob run was terminated.\n");*/
    if(!terminated) {           /* die quietly and quickly */
        if(options & GLOBRUN_PID)
            printf ("GLOB DONE=%d\n", (int)getpid ());
    }
    if(options & GLOBBER_VERBOSE) {
        fprintf (stdout, "fgr search complete!\n");
    }
    fflush (NULL);
    globber_destroy (object);
    exit (0);
}
